### PR TITLE
[FLINK-25528][state-processor-api] state processor api do not support increment checkpoint

### PR DESCRIPTION
**Pull request by liufangqi**
## What is the purpose of the change

Make state-processor-api support increment checkpoint write.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java:61]: I think you should make the checkpoint type configurable.
Moreover, I think the correct checkpoint type here should be `CheckpointType.FULL_CHECKPOINT`, please refer [FLINK-25192] for more details.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for creating the PR. Please take a look at my comments.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java
Comment (liufangqi) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java:61]: THX for your comments. I will make it configurable. 
And CheckpointType.FULL_CHECKPOINT is clearer type in new flink version.

**Comment by liufangqi** [comment]
@Myasuka Pushed a new commit. Could you help to review it when you are free? Just take your time, ping me if need.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java:57]: Generally, I think we should limit the snapshot type as full snapshot or savepoint instead of any kind of snapshot type.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the update. This PR deserves to add a test.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java
Comment (liufangqi) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java:57]: Good catch, I will add the limit and the unit test. THX!

**Comment by liufangqi** [comment]
@Myasuka Updated the code. Just take your time to review it again when you are free.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java
Comment (Myasuka) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/output/SnapshotUtils.java:59]: After re-think current stats in Flink, I currently prefer we could only support native and canonical savepoint. You can try to use native savepoint to see whether it could meet your previous requestls. The native savepoint is much close to full checkpoint and it could also support relocating the checkpoint data to other place without modifying the metadata.

**Review by Myasuka — COMMENTED** [review]
Thanks for the update, I think you could try Flink-1.15's native savepoint after FLINK-25276.

**Comment by liufangqi** [comment]
@Myasuka Agree the idea that only support Savepoint cause flink support the native savepoint in the new version. Updated the code that support switching the savepoint format type. Could you help review it again?

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java:56]: How about introduce a `@Before` method to clear the `ACTUAL_ORDER_TRACKING`?

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java:66]: I don't think this test actually test the snapshot type. I think we can recored the `snapshotType` in `LifecycleOperator#snapshotState` method, and then verify that we have passed the native savepoint snapshot type.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the update, I think the unit test stiil needs to be improved.

**Comment by liufangqi** [comment]
@Myasuka THX for your reply. You can see the update code in the new commit. I'm not sure that I got your meaning. Ping me if i am wrong.

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (liufangqi) [reply flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java]: It can make code reuse.

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java:81]: I think we can reduce duplicate code to make one private method `testSnapshotUtilsLifecycle(SavepointFormatType savepointFormatType)` and let two public test mthod just call it via `testSnapshotUtilsLifecycle(SavepointFormatType.DEFAULT)` and `testSnapshotUtilsLifecycle(SavepointFormatType.NATIVE)`.

By doing so, we can also drop the `initializeVariable` method.

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java:59]: I don't think we need a list to hold the snapshot type. Moreover, the expected snapshot type is what we passed in, an no need to introduce another list of `EXPECTED_SNAPSHOT_TYPE`.

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for the updating.
Please take a look at my comments.

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (Myasuka) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java:61]: In general, we should only use one unit test case for one specific test purpose. It's better to split this test case into two cases, one is testing `SavepointFormatType.DEFAULT` and another is for `SavepointFormatType.NATIVE`.

Moreover, I also suggest to test `SavepointFormatType.CANONICAL` instead of `SavepointFormatType.DEFAULT` in case we might change the `SavepointFormatType.DEFAULT` to `SavepointFormatType.NATIVE` in the future.

**Review by Myasuka — COMMENTED** [review]
I think this PR is close to merge, please take a look at my comments.

Before merging, could you also verify this patch locally to see whether the generated savepoint meets your request?

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java
Comment (liufangqi) [reply flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/output/SnapshotUtilsTest.java]: THX for your remind, actually I have already tested the generated savepoint locally and I tested it again just. The result met my request perfectly.
And I split the test case in the new commit, could you help take a glance?

**Comment by Myasuka** [comment]
The CI is still broken due to `KafkaSinkITCase.testAbortTransactionsAfterScaleInBeforeFirstCheckpoint`, would please check whther this is related to your changes @liufangqi 

If not, you can rebase with latest Flink master branch and then force push your `state-processor-api-support-checkpoint` branch,

**Comment by liufangqi** [comment]
@Myasuka THX for your remind.
> The CI is still broken due to `KafkaSinkITCase.testAbortTransactionsAfterScaleInBeforeFirstCheckpoint`, would please check whther this is related to your changes @liufangqi
> 
I have tested it locally, there is no problem.

> If not, you can rebase with latest Flink master branch and then force push your `state-processor-api-support-checkpoint` branch,

I have done the rebase job, the CI pass now. Could you help review it again?

**Review by Myasuka — APPROVED** [review]
LGTM. Thanks for the contribution!

Since the `release-1.15` branch has not been cut out and the feature freeze of `flink-1.15` has paased, I will merge this PR to master branch after `release-1.15` branch cut out.

**Pull request merged**
